### PR TITLE
importccl: fail IMPORT INTO on concurrent type change

### PR DESCRIPTION
**Pull request by adityamaru**
This change adds logic to check that the type descriptors
referenced by the table being imported into have not
undergone any changes (adding/dropping values) during the
the course of the import job execution. This is done by
matching the current type desc version against the version
on the type desc that was cached in the import job during
planning. If there is a mismatch, we should fail the job
since the data could be referencing a value that has been
dropped, leading to corrupt kv entries.

Informs: #61133

Release note (bug fix): add protection to import into to
guard against concurrent type changes on UDTs referenced by
the target table.

Release justification: fixes for high-priority or high-severity bugs in existing functionality

### CODE REVIEW

**Comment by adityamaru** [comment]
This is the first of a few PRs that gets us UDT support for `IMPORT INTO`. A follow-up will teach default expr + computed columns about UDTs.

**Comment by adityamaru** [comment]
@arulajmani / @ajwerner I was pleasantly surprised when my test for dropping an enum value mid import was failing with `could not validate enum value removal for "hi": relation "t" is offline: importing`. I think this sort of makes the second commit in this PR unnecessary? In fact, we would now fail the import job even if the enum is renamed, or grows which seems bad.

Unless y'all think otherwise I'll revert the second commit and only keep the tests testing that a type change during a running import behaves as expected.

**Review by ajwerner — COMMENTED** [review]
I think the concurrent changes we need to watch out for are type renames and enum member renames rather than value drops. Imagine that we swapped the names of two values. `ALTER TYPE typ RENAME VALUE a to c; ALTER TYPE typ RENAME VALUE b TO a; ALTER TYPE RENAME VALUE c TO b`. That'd be potentially bad as it might not be deterministic in the face of restarts. Same goes for the renaming of the type itself. 

I'm a bit worried about the fact that you can pause and resume and in the meantime change things. I suspect that that also is not safe. Perhaps what we need to do is look at the modification time on the descriptors to decide whether anything has changed. This is sort of a bummer because it means that grant/revoke to the type would trash your import. 

In summary, this is all pretty sad. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-:-MiWqg5uBgMvOrsiX-BV:b7yr7qh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @dt)

**Comment by adityamaru** [comment]
Yep, this all makes sense, thanks. We discussed this in our team meeting and feel like we should err on the side of caution and just fail the import if the modification time changes. This is probably an unnecessarily aggressive hammer but it doesn't seem unreasonable to tell users not to change their types during the import, at least for 21.2. We're also hoping we can hand this method over to SQL Schema to refine since they have a better understanding of what should/should not fail the import.
There is some urgency to power forward so that we can get to #61133 and https://github.com/cockroachdb/cockroach/issues/69616 but I'm happy to iterate on this.

**Comment by ajwerner** [comment]
> We're also hoping we can hand this method over to SQL Schema to refine since they have a better understanding of what should/should not fail the import.

Ack. If there's need for work from SQL Schema in 22.1, please make sure the issues are properly filed, tracked, and raised in the planning process so we can be sure that we get to it. In the meantime, the modification time approach seems fine.

**Review by arulajmani — COMMENTED** [review]
> Reviewed 4 of 4 files at r3, 1 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69674)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, and @dt)

___
*[-- commits, line 21 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib6J6O2c-izE1Usivy):*
Should this say by comparing modification time now?
___
*[-- commits, line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib6UhZ2q8VAmUkq-eH):*
This commit message could use an update now that we've discovered dropping enum values is no longer the problem here. 
___
*[pkg/ccl/importccl/import_stmt.go, line 294 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib3NiM0-HsRHIUxE5V:-Mib3NiM0-HsRHIUxE5W:b-uzpmwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/066be0bae6c7c5250ec878349181c4f4d401a36b/pkg/ccl/importccl/import_stmt.go#L294)):*
> ```Go
> func resolveUDTsUsedByImportInto(
> 	ctx context.Context, p sql.PlanHookState, table *tabledesc.Mutable,
> ) ([]*typedesc.Mutable, error) {
> ```

Instead of returning `typedesc.Mutable` here you could return `catalog.TypeDescriptor` instead. We're not really mutating the type descriptor and that interface gives you access to the underlying proto that you need below anyway. 
___
*[pkg/ccl/importccl/import_stmt.go, line 310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib34y730PrZC6ype0N:-Mib34y85OoRF76NdkKj:b-tbg8rd) ([raw file](https://github.com/cockroachdb/cockroach/blob/066be0bae6c7c5250ec878349181c4f4d401a36b/pkg/ccl/importccl/import_stmt.go#L310)):*
> ```Go
> 		typeIDs, _, err := table.GetAllReferencedTypeIDs(dbDesc,
> 			func(id descpb.ID) (catalog.TypeDescriptor, error) {
> 				mutDesc, err := descriptors.GetMutableTypeByID(ctx, txn, id, tree.ObjectLookupFlags{
> ```

nit: Let's use GetImmutableTypeByID instead? 
___
*[pkg/ccl/importccl/import_stmt.go, line 326 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib3bF93bKs6xVjZD-R:-Mib3bF93bKs6xVjZD-S:bcnmcv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/066be0bae6c7c5250ec878349181c4f4d401a36b/pkg/ccl/importccl/import_stmt.go#L326)):*
> ```Go
> 
> 		for _, typeID := range typeIDs {
> 			mutDesc, err := descriptors.GetMutableTypeByID(ctx, txn, typeID, tree.ObjectLookupFlags{
> ```

If you change the return type of this function you can use `GetImmutable...` here as well. 
___
*[pkg/ccl/importccl/import_stmt.go, line 866 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib4A03Aq55KJiO4RoQ:-Mib4A03Aq55KJiO4RoR:bclmaic) ([raw file](https://github.com/cockroachdb/cockroach/blob/066be0bae6c7c5250ec878349181c4f4d401a36b/pkg/ccl/importccl/import_stmt.go#L866)):*
> ```Go
> 				}
> 				if len(typeDescs) > 0 {
> 					typeDetails = make([]jobspb.ImportDetails_Type, 0)
> ```

nit: you could pass in `len(typeDescs)` for the capacity here. 
___
*[pkg/ccl/importccl/import_stmt.go, line 2286 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib4h7M9V2bCRdGL8vk:-Mib4h7M9V2bCRdGL8vl:b-ldumed) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbbae90ff86e7ef56f20e5d7143bad6bdc10c9a/pkg/ccl/importccl/import_stmt.go#L2286)):*
> ```Go
> // table being imported into have been modified since they were read during
> // import planning. If they have, it may be unsafe to continue with the import
> // since we could be ingesting data that is no longer valid for the type.
> ```

Might be worth extending this comment with an example of a change to the underlying type which can make the import unsafe. 
___
*[pkg/ccl/importccl/import_stmt.go, line 2301 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib5Hl_Ek9jwoH-JLDg:-Mib5Hl_Ek9jwoH-JLDh:b-p74gp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbbae90ff86e7ef56f20e5d7143bad6bdc10c9a/pkg/ccl/importccl/import_stmt.go#L2301)):*
> ```Go
> 	return execCfg.DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
> 		for _, savedTypeDesc := range details.Types {
> 			typeDesc, err := catalogkv.MustGetTypeDescByID(ctx, txn, execCfg.Codec,
> ```

Is there a reason to use a `catalogkv` method here instead of going through the descsCollection? Could we instead use `sql.DescsTxn` here? I don't think there is a correctness argument for my suggestion, just that it'd be nice to avoid this lower level method if we can. 
___
*[pkg/ccl/importccl/import_stmt_test.go, line 7089 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib75hP3XZJWy8pZo3Q:-Mib75hP3XZJWy8pZo3R:b-8qoyld) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbbae90ff86e7ef56f20e5d7143bad6bdc10c9a/pkg/ccl/importccl/import_stmt_test.go#L7089)):*
> ```Go
> }
> 
> func TestUDTChangeDuringImport(t *testing.T) {
> ```

I don't think this is related to this PR, but I was talking to @adityamaru about general IMPORT INTO <-> UDTs testing and it looks like we're lacking a test for `IMPORT INTO` if an enum value doesn't exist on the type descriptor but is in the import file. Might be worth adding such a thing in a followup/separate commit. 
___
*[pkg/ccl/importccl/import_stmt_test.go, line 7134 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-MibCnsD7Cy3LTrF0Qtn:-MibCnsD7Cy3LTrF0Qto:b-y4zl83) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbbae90ff86e7ef56f20e5d7143bad6bdc10c9a/pkg/ccl/importccl/import_stmt_test.go#L7134)):*
> ```Go
> 			"ALTER TYPE d.greeting DROP VALUE 'howdy';",
> 			"could not validate enum value removal",
> 			true,
> ```

Might be worth adding a comment here why the import fails even though the type change is failing

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69674)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @dt)

___
*[-- commits, line 21 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib6J6O2c-izE1Usivy):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should this say by comparing modification time now?
</blockquote></details>

yup, changed.
___
*[-- commits, line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib6UhZ2q8VAmUkq-eH):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This commit message could use an update now that we've discovered dropping enum values is no longer the problem here. 
</blockquote></details>

Changed to rename since that is an understood example of our problem.
___
*[pkg/ccl/importccl/import_stmt.go, line 294 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib3NiM0-HsRHIUxE5V:-MibZEaRE2J0VTrMiL2j:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/066be0bae6c7c5250ec878349181c4f4d401a36b/pkg/ccl/importccl/import_stmt.go#L294)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Instead of returning `typedesc.Mutable` here you could return `catalog.TypeDescriptor` instead. We're not really mutating the type descriptor and that interface gives you access to the underlying proto that you need below anyway. 
</blockquote></details>

done.
___
*[pkg/ccl/importccl/import_stmt.go, line 310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib34y730PrZC6ype0N:-MibZFWa95TJzq9yBPMb:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/066be0bae6c7c5250ec878349181c4f4d401a36b/pkg/ccl/importccl/import_stmt.go#L310)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Let's use GetImmutableTypeByID instead? 
</blockquote></details>

done.
___
*[pkg/ccl/importccl/import_stmt.go, line 326 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib3bF93bKs6xVjZD-R:-MibZr3YE8DJXNn5HIOU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/066be0bae6c7c5250ec878349181c4f4d401a36b/pkg/ccl/importccl/import_stmt.go#L326)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

If you change the return type of this function you can use `GetImmutable...` here as well. 
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/import_stmt.go, line 866 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib4A03Aq55KJiO4RoQ:-MibZt_u-5JqPYmt3Zna:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/066be0bae6c7c5250ec878349181c4f4d401a36b/pkg/ccl/importccl/import_stmt.go#L866)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: you could pass in `len(typeDescs)` for the capacity here. 
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/import_stmt.go, line 2286 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib4h7M9V2bCRdGL8vk:-MibiqKv6-AewqqXslfM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbbae90ff86e7ef56f20e5d7143bad6bdc10c9a/pkg/ccl/importccl/import_stmt.go#L2286)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Might be worth extending this comment with an example of a change to the underlying type which can make the import unsafe. 
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/import_stmt.go, line 2301 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib5Hl_Ek9jwoH-JLDg:-MibkMwA93m4E2Ozuv3k:b7svfwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbbae90ff86e7ef56f20e5d7143bad6bdc10c9a/pkg/ccl/importccl/import_stmt.go#L2301)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is there a reason to use a `catalogkv` method here instead of going through the descsCollection? Could we instead use `sql.DescsTxn` here? I don't think there is a correctness argument for my suggestion, just that it'd be nice to avoid this lower level method if we can. 
</blockquote></details>

none at all, switched to sql.Descs.
___
*[pkg/ccl/importccl/import_stmt_test.go, line 7089 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib75hP3XZJWy8pZo3Q:-Mibs4AC7vMTONz1R-3K:b-fni4i5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbbae90ff86e7ef56f20e5d7143bad6bdc10c9a/pkg/ccl/importccl/import_stmt_test.go#L7089)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't think this is related to this PR, but I was talking to @adityamaru about general IMPORT INTO <-> UDTs testing and it looks like we're lacking a test for `IMPORT INTO` if an enum value doesn't exist on the type descriptor but is in the import file. Might be worth adding such a thing in a followup/separate commit. 
</blockquote></details>

Added to `TestImportUserDefinedTypes`!
___
*[pkg/ccl/importccl/import_stmt_test.go, line 7134 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-MibCnsD7Cy3LTrF0Qtn:-Mibb0tN3SpjAfW4np48:b-gwld75) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbbae90ff86e7ef56f20e5d7143bad6bdc10c9a/pkg/ccl/importccl/import_stmt_test.go#L7134)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Might be worth adding a comment here why the import fails even though the type change is failing 
</blockquote></details>

Done. I haven't really dug into why the mod time actually changes even though we eventually forbid the enum drop.

**Review by arulajmani — APPROVED** [review]
:lgtm: 

> Reviewed 4 of 4 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-:-Mic0nVeD0Mck7l2pzfe:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, and @dt)

___
*[pkg/ccl/importccl/import_stmt_test.go, line 7089 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-Mib75hP3XZJWy8pZo3Q:-Mic-gfW20u6j1a3q_O0:b-ecpsw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbbae90ff86e7ef56f20e5d7143bad6bdc10c9a/pkg/ccl/importccl/import_stmt_test.go#L7089)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

Added to `TestImportUserDefinedTypes`!
</blockquote></details>

:100: :100: 
___
*[pkg/ccl/importccl/import_stmt_test.go, line 7134 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69674#-MibCnsD7Cy3LTrF0Qtn:-Mic-j96-R19EY6T_EV3:bp3m4c5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbbae90ff86e7ef56f20e5d7143bad6bdc10c9a/pkg/ccl/importccl/import_stmt_test.go#L7134)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

Done. I haven't really dug into why the mod time actually changes even though we eventually forbid the enum drop.
</blockquote></details>

I think it's because the OFFLINE error that you're expecting comes from the async portion of the schema change. The user transaction moves the enum member from `PUBLIC` to `READ ONLY` which results in an update to the modification time. The error you're expecting happens when we validate the enum member usages -- we can't for this table because it is offline. Validation fails and the enum member is transitioned back to `PUBLIC` from `READ ONLY`, resulting in another update to the modification time. 

This might be too much detail for your comment, but you're better with words than me. :P

**Comment by adityamaru** [comment]
@dt can i get a review on the import stuff, thanks!

**Review by dt — APPROVED** [review]

**Comment by adityamaru** [comment]
I really don't know why this bazel build is timing out under testrace while CI is passing. On average it seems like tests in `importccl` are taking much longer on bazel race runs than in CI. I'll keep investigating but given the timing of this and follow-up PRs I'm inclined to check this in.

**Comment by adityamaru** [comment]
bors r=arulajmani,dt
